agent/acs/handler/payload_responder.go

```diff
@@ -0,0 +1,289 @@
+	if !stoppedTasksAddedOK || !newTasksAddedOK {
+		allTasksOK = false
+	}
+
+	// Construct a slice with credentials acks from all tasks.
+	credentialsAcks := append(stoppedTasksCredentialsAcks, newTasksCredentialsAcks...)
+	return credentialsAcks, allTasksOK
+}
+
+// handleUnrecognizedTask handles unrecognized tasks by sending 'stopped' with
+// a suitable reason to the backend.
+func (tmHandler *taskMessageHandler) handleUnrecognizedTask(task *ecsacs.Task, err error,
+	payload *ecsacs.PayloadMessage) {
+	logger.Warn("Received unexpected ACS message", logger.Fields{
+		loggerfield.MessageID: aws.StringValue(payload.MessageId),
+		loggerfield.TaskARN:   aws.StringValue(task.Arn),
+		loggerfield.Error:     err,
+	})
+
+	if aws.StringValue(task.Arn) == "" {
+		logger.Critical("Received task with no ARN for payload message", logger.Fields{
+			loggerfield.MessageID: aws.StringValue(payload.MessageId),
+		})
+		return
+	}
+
+	// Only need to stop the task; it brings down the containers too.
+	taskEvent := api.TaskStateChange{
+		TaskARN: *task.Arn,
+		Status:  apitaskstatus.TaskStopped,
+		Reason:  UnrecognizedTaskError{err}.Error(),
+		// The real task cannot be extracted from payload message, so we send an empty task.
+		// This is necessary because the task handler will not send an event whose
+		// Task is nil.
+		Task: &apitask.Task{},
+	}
+
+	tmHandler.taskHandler.AddStateChangeEvent(taskEvent, tmHandler.ecsClient)
+}
+
+// addTasks adds the tasks to the task engine based on the skipAddTask condition.
+// This is used to add non-stopped tasks before adding stopped tasks.
+func (tmHandler *taskMessageHandler) addTasks(payload *ecsacs.PayloadMessage, tasks []*apitask.Task,
+	skipAddTask skipAddTaskComparatorFunc) ([]*ecsacs.IAMRoleCredentialsAckRequest, bool) {
+	allTasksOK := true
+	var credentialsAcks []*ecsacs.IAMRoleCredentialsAckRequest
+	for _, task := range tasks {
+		if skipAddTask(task.GetDesiredStatus()) {
+			continue
+		}
+		tmHandler.taskEngine.AddTask(task)
+		// Only need to save task to DB when its desired status is RUNNING (i.e. this is a new task that we are going
+		// to manage). When its desired status is STOPPED, the task is already in the DB and the desired status change
+		// will be saved by task manager.
+		if task.GetDesiredStatus() == apitaskstatus.TaskRunning {
+			err := tmHandler.dataClient.SaveTask(task)
+			if err != nil {
+				logger.Error("Failed to save data for task", logger.Fields{
+					loggerfield.TaskARN: task.Arn,
+					loggerfield.Error:   err,
+				})
+				allTasksOK = false
+			}
+		}
+
+		ackCredentials := func(id string, description string) {
+			ack, err := tmHandler.ackCredentials(payload.MessageId, id)
+			if err != nil {
+				allTasksOK = false
+				logger.Error(fmt.Sprintf("Failed to acknowledge %s credentials for task",
+					description), logger.Fields{
+					"task":            task.String(),
```

Comment:
You can use the logger field: TaskID instead of "task".

Reply:
I believe we need should use "task" here since `task.String()` doesn't return just a task ID, but it returns a string representation of the task (which includes the task ARN).

Reply:
Agreed.TaskID might get a bit confusing, even though it actually corresponds to "task" in constants file. Thanks.